haystack/document_stores/base.py

```diff
@@ -51,7 +51,7 @@ def query(self, sparql_query: str, index: Optional[str] = None, headers: Optiona
         raise NotImplementedError
 
 
-class BaseDocumentStore(BaseComponent):
+class BaseDocumentStore(BaseComponent, ABC):
```

Comment:
ABC is quite special AFAIK: subclasses of ABC are not ABC themselves. So if I want to have a hierarchy of abstract classes (like here), they must all explicitly inherit from ABC. I don't think there's a way around this. I need `BaseDocumentStore` to be abstract, and this seems to be the only way to do this.

That's just my understanding of how ABC work though. If that's not true please correct me :+1:

In addition, even if this was not the case, I would not rely on the presence of `@abstractmethod` decorators to tell if a class is abstract or not. That's a fundamental property of the class and I'd rather be explicit even if it was not a necessity...

Reply:
Nice topic. 🙂 Now that BaseComponent inherits from ABC, all subclasses of BaseComponent will implicitly also inherit from ABC. I double-checked with the code example from [here](https://stackoverflow.com/a/51232156). Explicitly inheriting from ABC is thus not necessary.
I agree with the point that even if `BaseDocumentStore` explicitly inherited from ABC it could be instantiated if it had no `@abstractmethod` declaration and if it implemented all `@abstractmethod`s of `BaseComponent`, namely `run`. So just from the inheritance of ABC one can't tell whether the inheriting class can be instantiated or not. Therefore, I would argue to use only `class BaseDocumentStore(BaseComponent):` without explicit inheritance from ABC.

Reply:
Wow, thanks! I got it completely wrong. I'll remove the redundant ABCs :+1: